includes/polling/ports.inc.php

```diff
@@ -334,7 +334,23 @@
             $port_ent_to_if = [];
             foreach ($ifTable_ifDescr as $if_index => $if_descr) {
                 if (preg_match('/^[[:alpha:]]+ethernet([0-9\/.]+)$/i', $if_descr['ifDescr'], $matches)) {
-                    $port_ent_to_if[str_replace('/', '.', $matches[1])] = ['portIfIndex' => $if_index];
+                    /*
+                    The ...EthernetX/Y/Z SNMP entries on Catalyst 9x00 iosxe
+                    are cpeExtStuff.X.Z instead of cpeExtStuff.X.Y.Z
+                    Creating both:
+                      $port_ent_to_if[slot.subslot.port]
+                      $port_ent_to_if[slot.port]
+                    will find the PoE stats for Cat9K and will also match for
+                    devices that do include the subslot (for subslot == 0)
+                    */
+                    $iface_parts = explode('/', $matches[1]);
+                    while (1) {
```

Comment:
while (1) needs to change, you could honestly update the regex above and drop most of this code.

Reply:
I left the regexp from the original code as we still need to match ethernet interfaces and capture the slot[/subslot]/port part (to be able replace '/' by '.' later.)
Capturing repeated subgroups in PHP's PCRE would make for a hard to read regexp. Maybe that was your intention ? I'm not quite sure what you meant by 'update the regexp'.

We also still need to add both slot.subslot.port and slot.port to $port_ent_to_if, i.e. two assignments (at least). Otherwise we'd break the poller for non-cat9K devices.
We don't want to strip the subslot element if it's non-zero, i.e. we need to check that part of the interface notation.
Maybe your assumption is that there will always be either slot/subslot/port or slot/port interfaces, and not e.g. slot/subslot/subsubslot/port. The PR code doesn't make any assumptions.

Reply:
The while (1) is in effect a do..while, preventing a duplication of the assignment statements.
I'll see if  a variation on that theme with a for loop or some such is less offensive to the eyes :)

Reply:
```php
 if (preg_match('/^[a-z]+ethernet(\d+)(?:\/(\d+))(?:\/(\d+))?$/i', $if_descr['ifDescr'], $matches)) {
port_ent_to_if[$matches[1] . '.' . $matches[3] ?: $matches[2]] = ['portIfIndex' => $if_index];
}
```

Should be it.

Reply:
> Should be it. Specifically only matches XethernetX/X or XethernetX/X/X
> If cpeExtPsePortEntry can be indexed differently, this code is probably slightly wrong.

That's indeed the catch: I fully expect there to be an existing ios(xe) device that indexes XethernetA/B/C in cpeExtPsePortEntry with a.b.c. even with B == 0. Breaking the poller is bad, hence the code to add both A.B.C. and A.C to $port_ent_to_if, for arbitrary 'depths' of interface names.

However, if we're working with the assumption that there will only ever be XethernetX/X or XethernetX/X/X (i.e. two or three parts notation), we can refactor the PR as two assignment statements, like so:

```php
if (preg_match('/^\w+ethernet(\d+)\/(\d+)(?:\/(\d+))?$/i', $if_descr['ifDescr'], $matches)) {
    $port_ent_to_if[$matches[1] . '.' . $matches[2]] =
        $port_ent_to_if[$matches[1] . '.' . ($matches[3] ?: $matches[2])] =
            ['portIfIndex' => $if_index];
}
```
What do you think ?
(or split it up over two individual assignment statements)

Reply:
> Should be it. Specifically only matches XethernetX/X or XethernetX/X/X
> If cpeExtPsePortEntry can be indexed differently, this code is probably slightly wrong.

That's indeed the catch: I fully expect there to be an existing ios(xe) device that indexes XethernetA/B/C in cpeExtPsePortEntry with a.b.c. even with B == 0. Breaking the poller is bad, hence the code to add both A.B.C. and A.C to $port_ent_to_if, for arbitrary 'depths' of interface names.

However, if we're working with the assumption that there will only ever be XethernetX/X or XethernetX/X/X (i.e. two or three parts notation), we can refactor the PR as two assignment statements, like so:

```php
if (preg_match('/^\w+ethernet(\d+)\/(\d+)(?:\/(\d+))?$/i', $if_descr['ifDescr'], $matches)) {
    foreach (array(0=>0) + ($matches[3] ? array(1=>2):array()) as $idx) {
        unset($matches[$idx]);
        $port_ent_to_if[implode('.', $matches)] = ['portIfIndex' => $if_index];
    }
}
```
What do you think ?

Reply:
Why do you expect that? Generally the table indexes are fixed.  I don't have any sample data to l look at so I can't confirm.  (btw you should add test data to this PR.

Reply:
why the loop?
just add two entries into the $port_ent_to_if array

Reply:
Added tests. I've check with ```./lnms dev:check -o iosxe_c9400``` , they seem to pass.
Not sure about the 2nd ASA device in the JSON, did snmpsim inject it ?

Reply:
codeclimate once again complains about "Avoid deeply nested control flow statements". Can't be fixed without drastic rewriting all of the code.

Node analysis timed out "The command produced no output for 180 seconds. You can increase this timeout by defining the "idle_timeout" modifier.". Not something I can fix :)